chainerx_cc/chainerx/cuda/cuda_device/indexing.cu

```diff
@@ -267,6 +267,32 @@ public:
 
 CHAINERX_CUDA_REGISTER_KERNEL(AddAtKernel, CudaAddAtKernel);
 
+template <typename T>
+struct WhereImpl {
+    using CudaType = cuda_internal::DataType<T>;
+    __device__ void operator()(int64_t /*i*/, bool condition, CudaType x, CudaType y, CudaType& out) { out = condition ? x : y; }
+};
+
+class CudaWhereKernel : public WhereKernel {
+public:
+    void Call(const Array& condition, const Array& x, const Array& y, const Array& out) override {
+        Device& device = x.device();
+        device.CheckDevicesCompatible(condition, x, y, out);
+        Dtype dtype = std::max(x.dtype(), y.dtype());
+        const Array& x_cast = x.dtype() == dtype ? x : x.AsType(dtype);
+        const Array& y_cast = y.dtype() == dtype ? y : y.AsType(dtype);
+        const Array& out_cast = out.dtype() == dtype ? out : out.AsType(dtype);
```

Comment:
The output will not be updated when `out.dtype == dtype` is false.

Reply:
Actually this case doesn't happen as we make output, the same-type as `x and y`. Also note that, we don't allow `x` and `y` to be of different `dtypes` similar to onnx description. However numpy promotes the lower dtype. Should we also do the same?

Reply:
For consistency with other functions, this function should support mixed dtype. For example, `chainerx::Concatenate` supports mixed dtype whereas onnx's `Concat` requires the same types.

Reply:
Oh. Done.